File: DMS_InstDirScanner/clsMgrSettings.vb
'*********************************************************************************************************
' Written by Dave Clark for the US Department of Energy
' Pacific Northwest National Laboratory, Richland, WA
' Copyright 2007, Battelle Memorial Institute
' Created 07/27/2009
'
'*********************************************************************************************************
Imports System.Xml

#Region "Interfaces"
Public Interface IMgrParams
    Function GetParam(ItemKey As String) As String
    Function GetParam(ItemKey As String, valueIfMissing As String) As String
    Function GetParam(ItemKey As String, valueIfMissing As Integer) As Integer
    Sub SetParam(ItemKey As String, ItemValue As String)
End Interface
#End Region

''' <summary>
''' Class for loading, storing and accessing manager parameters.
''' </summary>
''' <remarks>
''' Loads initial settings from local config file, then checks to see if remainder of settings should be
''' loaded or manager set to inactive. If manager active, retrieves remainder of settings from manager
''' parameters database.
''' </remarks>
Public Class clsMgrSettings
    Implements IMgrParams

#Region "Module variables"
    Private m_MgrParams As Dictionary(Of String, String)
    Private m_ErrMsg As String = ""
#End Region

#Region "Properties"
    Public ReadOnly Property ErrMsg As String
        Get
            Return m_ErrMsg
        End Get
    End Property
#End Region

#Region "Methods"
    ''' <summary>
    ''' Constructor
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub New()

        If Not LoadSettings() Then
            Throw New ApplicationException("Unable to initialize manager settings class")
        End If

    End Sub

    ''' <summary>
    ''' Loads manager settings from config file and database
    ''' </summary>
    ''' <returns>True if successful; False on error</returns>
    ''' <remarks></remarks>
    Public Function LoadSettings() As Boolean

        m_ErrMsg = ""

        'Get settings from config file
        m_MgrParams = LoadMgrSettingsFromFile()

        'Test the settings retrieved from the config file
        If Not CheckInitialSettings(m_MgrParams) Then
            'Error logging was already handled by CheckInitialSettings
            Return False
        End If

        'Determine if manager is deactivated locally
        If Not CBool(m_MgrParams("MgrActive_Local")) Then
            LogWarning("Manager deactivated locally")
            m_ErrMsg = "Manager deactivated locally"
            Return False
        End If

        'Get remaining settings from database
        If Not LoadMgrSettingsFromDB(m_MgrParams) Then
            'Error logging handled by LoadMgrSettingsFromDB
            Return False
        End If

        'No problems found
        Return True

    End Function

    ''' <summary>
    ''' Loads the initial settings from application config file
    ''' </summary>
    ''' <returns>String dictionary containing initial settings if suceessful; NOTHING on error</returns>
    ''' <remarks></remarks>
    Private Function LoadMgrSettingsFromFile() As Dictionary(Of String, String)

        'Load initial settings into string dictionary for return
        Dim mgrParams As New Dictionary(Of String, String)(StringComparer.OrdinalIgnoreCase)

        My.Settings.Reload()
        'Manager config db connection string
        mgrParams.Add("MgrCnfgDbConnectStr", My.Settings.MgrCnfgDbConnectStr)

        'Manager active flag
        mgrParams.Add("MgrActive_Local", My.Settings.MgrActive_Local.ToString())

        'Manager name
        mgrParams.Add("MgrName", My.Settings.MgrName)

        'Default settings in use flag
        mgrParams.Add("UsingDefaults", My.Settings.UsingDefaults.ToString())

        Return mgrParams

    End Function

    ''' <summary>
    ''' Tests initial settings retrieved from config file
    ''' </summary>
    ''' <param name="mgrParams"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Private Function CheckInitialSettings(mgrParams As IReadOnlyDictionary(Of String, String)) As Boolean

        'Verify manager settings dictionary exists
        If mgrParams Is Nothing Then
            LogError("clsMgrSettings.CheckInitialSettings(); Manager parameters dictionary is null")
            Return False
        End If

        'Verify intact config file was found
        Dim strUsingDefaults = ""
        If Not mgrParams.TryGetValue("UsingDefaults", strUsingDefaults) Then
            LogError("clsMgrSettings.CheckInitialSettings(); Manager parameter 'UsingDefaults' is not defined")
            Return False
        End If

        Dim usingDefaults As Boolean
        If Not Boolean.TryParse(strUsingDefaults, usingDefaults) Then
            LogError("clsMgrSettings.CheckInitialSettings(); Manager parameter 'UsingDefaults' must be True or False, not " & strUsingDefaults)
            Return False
        End If

        If usingDefaults Then
            LogError("clsMgrSettings.CheckInitialSettings(); Config file problem: default settings being used (UsingDefaults is true)")
            Return False
        End If

        'No problems found
        Return True

    End Function

    ''' <summary>
    ''' Gets remaining manager config settings from config database;
    ''' Overload to use module-level string dictionary when calling from external method
    ''' </summary>
    ''' <returns>True for success; False for error</returns>
    ''' <remarks></remarks>
    Public Overloads Function LoadMgrSettingsFromDB() As Boolean

        Return LoadMgrSettingsFromDB(m_MgrParams)

    End Function


    ''' <summary>
    ''' Gets remaining manager config settings from config database
    ''' </summary>
    ''' <param name="mgrSettings">String dictionary containing parameters that have been loaded so far</param>
    ''' <returns>True for success; False for error</returns>
    ''' <remarks></remarks>
    Public Overloads Function LoadMgrSettingsFromDB(mgrSettings As Dictionary(Of String, String)) As Boolean

        'Requests job parameters from database. Input string specifies view to use. Performs retries if necessary.

        Dim mgrName = m_MgrParams("MgrName")

        Dim columns = New List(Of String) From {
                "ParameterName",
                "ParameterValue"
                }

        Dim sqlStr As String =
                " SELECT " & String.Join(",", columns) &
                " FROM V_MgrParams " &
                " WHERE ManagerName = '" & mgrName & "'"

        Dim connectionString = mgrSettings("MgrCnfgDbConnectStr")
        If String.IsNullOrWhiteSpace(connectionString) Then
            LogError("Connection string is empty; cannot retrieve manager parameters")
            Return False
        End If

        Dim dbTools = New PRISM.clsDBTools(connectionString)
        AddHandler dbTools.ErrorEvent, AddressOf DBToolsErrorHandler
        AddHandler dbTools.WarningEvent, AddressOf DBToolsWarningHandler

        ' Get a table holding the parameters for one manager

        Dim lstResults As List(Of List(Of String)) = Nothing
        Dim success = dbTools.GetQueryResults(sqlStr, lstResults, "LoadMgrSettingsFromDB")

        If Not success Then
            LogError("Unable to retrieve manager settings from the database for manager " & mgrName)
            Return False
        End If

        'Verify at least one row returned
        If lstResults.Count < 1 Then
            LogError("clsMgrSettings.LoadMgrSettingsFromDB; manager settings not found in V_MgrParams for manager " & mgrName)
            Return False
        End If

        Dim colMapping = dbTools.GetColumnMapping(columns)

        'Fill a string dictionary with the manager parameters that have been found
        Try
            For Each result In lstResults

                Dim paramKey = dbTools.GetColumnValue(result, colMapping, "ParameterName")
                Dim paramVal = dbTools.GetColumnValue(result, colMapping, "ParameterValue")

                If m_MgrParams.ContainsKey(paramKey) Then
                    m_MgrParams(paramKey) = paramVal
                Else
                    m_MgrParams.Add(paramKey, paramVal)
                End If
            Next

            Return True
        Catch ex As Exception
            LogError("clsMgrSettings.LoadMgrSettingsFromDB; Exception filling string dictionary from table: " & ex.Message)
            Return False
        End Try

    End Function

    Private Sub LogError(message As String)
        PRISM.ConsoleMsgUtils.ShowError(message)
        clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogSystem, clsLogTools.LogLevels.ERROR, message)
    End Sub

    Private Sub LogWarning(message As String)
        PRISM.ConsoleMsgUtils.ShowWarning(message)
        clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, message)
    End Sub

    ''' <summary>
    ''' Gets a parameter from the parameters string dictionary
    ''' </summary>
    ''' <param name="ItemKey">Key name for item</param>
    ''' <returns>String value associated with specified key</returns>
    ''' <remarks>Returns Nothing if key isn't found</remarks>
    Public Function GetParam(ItemKey As String, valueIfMissing As String) As String Implements IMgrParams.GetParam

        If Not m_MgrParams.ContainsKey(ItemKey) Then
            Return valueIfMissing
        End If

        Return m_MgrParams.Item(ItemKey)

    End Function

    ''' <summary>
    ''' Gets a parameter from the parameters string dictionary
    ''' </summary>
    ''' <param name="ItemKey">Key name for item</param>
    ''' <returns>String value associated with specified key</returns>
    ''' <remarks>Returns Nothing if key isn't found</remarks>
    Public Function GetParam(ItemKey As String, valueIfMissing As Integer) As Integer Implements IMgrParams.GetParam

        If Not m_MgrParams.ContainsKey(ItemKey) Then
            Return valueIfMissing
        End If

        Dim strValue = m_MgrParams.Item(ItemKey)
        Dim value As Integer
        If Integer.TryParse(strValue, value) Then
            Return value
        End If

        Return valueIfMissing

    End Function

    ''' <summary>
    ''' Gets a parameter from the parameters string dictionary
    ''' </summary>
    ''' <param name="ItemKey">Key name for item</param>
    ''' <returns>String value associated with specified key</returns>
    ''' <remarks>Returns Nothing if key isn't found</remarks>
    Public Function GetParam(ItemKey As String) As String Implements IMgrParams.GetParam

        Return m_MgrParams.Item(ItemKey)

    End Function

    ''' <summary>
    ''' Sets a parameter in the parameters string dictionary
    ''' </summary>
    ''' <param name="ItemKey">Key name for the item</param>
    ''' <param name="ItemValue">Value to assign to the key</param>
    ''' <remarks></remarks>
    Public Sub SetParam(ItemKey As String, ItemValue As String) Implements IMgrParams.SetParam

        m_MgrParams.Item(ItemKey) = ItemValue

    End Sub

    ''' <summary>
    ''' Gets a collection representing all keys in the parameters string dictionary
    ''' </summary>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetAllKeys() As ICollection

        Return m_MgrParams.Keys

    End Function

    ''' <summary>
    ''' Writes specfied value to an application config file.
    ''' </summary>
    ''' <param name="Key">Name for parameter (case sensitive)</param>
    ''' <param name="Value">New value for parameter</param>
    ''' <returns>TRUE for success; FALSE for error (ErrMsg property contains reason)</returns>
    ''' <remarks>This bit of lunacy is needed because MS doesn't supply a means to write to an app config file</remarks>
    Public Function WriteConfigSetting(Key As String, Value As String) As Boolean

        m_ErrMsg = ""

        'Load the config document
        Dim MyDoc As XmlDocument = LoadConfigDocument()
        If MyDoc Is Nothing Then
            'Error message has already been produced by LoadConfigDocument
            Return False
        End If

        'Retrieve the settings node
        Dim MyNode As XmlNode = MyDoc.SelectSingleNode("//applicationSettings")

        If MyNode Is Nothing Then
            m_ErrMsg = "clsMgrSettings.WriteConfigSettings; appSettings node not found"
            Return False
        End If

        Try
            'Select the eleement containing the value for the specified key containing the key
            Dim MyElement = CType(MyNode.SelectSingleNode(String.Format("//setting[@name='{0}']/value", Key)), XmlElement)
            If MyElement IsNot Nothing Then
                'Set key to specified value
                MyElement.InnerText = Value
            Else
                'Key was not found
                m_ErrMsg = "clsMgrSettings.WriteConfigSettings; specified key not found: " & Key
                Return False
            End If
            MyDoc.Save(GetConfigFilePath())
            Return True
        Catch ex As Exception
            m_ErrMsg = "clsMgrSettings.WriteConfigSettings; Exception updating settings file: " & ex.Message
            Return False
        End Try

    End Function

    ''' <summary>
    ''' Loads an app config file for changing parameters
    ''' </summary>
    ''' <returns>App config file as an XML document if successful; NOTHING on failure</returns>
    ''' <remarks></remarks>
    Private Function LoadConfigDocument() As XmlDocument

        Try
            Dim myDoc = New XmlDocument
            myDoc.Load(GetConfigFilePath)
            Return myDoc
        Catch ex As Exception
            m_ErrMsg = "clsMgrSettings.LoadConfigDocument; Exception loading settings file: " & ex.Message
            Return Nothing
        End Try

    End Function

    ''' <summary>
    ''' Specifies the full name and path for the application config file
    ''' </summary>
    ''' <returns>String containing full name and path</returns>
    ''' <remarks></remarks>
    Private Function GetConfigFilePath() As String

        Return clsMainProcess.GetAppPath() & ".config"

    End Function

    ''' <summary>
    ''' Converts a database output object that could be dbNull to a string
    ''' </summary>
    ''' <param name="InpObj"></param>
    ''' <returns>String equivalent of object; empty string if object is dbNull</returns>
    ''' <remarks></remarks>
    Protected Function DbCStr(InpObj As Object) As String

        'If input object is DbNull, returns "", otherwise returns String representation of object
        If InpObj Is DBNull.Value Then
            Return ""
        Else
            Return CStr(InpObj)
        End If

    End Function

    Private Sub DBToolsErrorHandler(message As String, ex As Exception)
        LogError(message)
    End Sub

    Private Sub DBToolsWarningHandler(message As String)
        LogWarning(message)
    End Sub

#End Region

End Class

File: DMS_InstDirScanner/clsMgrSettings.cs

//*********************************************************************************************************
// Written by Dave Clark for the US Department of Energy
// Pacific Northwest National Laboratory, Richland, WA
// Copyright 2009, Battelle Memorial Institute
// Created 06/16/2009
//*********************************************************************************************************

using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using System.IO;
using PRISM;

namespace DMS_InstDirScanner
{
    /// <summary>
    /// Class for loading, storing and accessing manager parameters.
    /// </summary>
    /// <remarks>
    /// Loads initial settings from local config file, then checks to see if remainder of settings should be
    /// loaded or manager set to inactive. If manager active, retrieves remainder of settings from manager
    /// parameters database.
    /// </remarks>
    public class MgrSettings : EventNotifier
    {
        #region "Constants"

        /// <summary>
        /// Status message for when the manager is deactivated locally
        /// </summary>
        /// <remarks>Used when MgrActive_Local is False in AppName.exe.config</remarks>
        public const string DEACTIVATED_LOCALLY = "Manager deactivated locally";

        /// <summary>
        /// Manager parameter: config database connection string
        /// </summary>
        public const string MGR_PARAM_MGR_CFG_DB_CONN_STRING = "MgrCnfgDbConnectStr";
        /// <summary>
        /// Manager parameter: manager active
        /// </summary>
        /// <remarks>Defined in AppName.exe.config</remarks>
        public const string MGR_PARAM_MGR_ACTIVE_LOCAL = "MgrActive_Local";

        /// <summary>
        /// Manager parameter: manager name
        /// </summary>
        public const string MGR_PARAM_MGR_NAME = "MgrName";

        /// <summary>
        /// Manager parameter: using defaults flag
        /// </summary>
        public const string MGR_PARAM_USING_DEFAULTS = "UsingDefaults";

        #endregion

        #region "Properties"

        /// <summary>
        /// Error message
        /// </summary>
        public string ErrMsg { get; private set; } = "";

        /// <summary>
        /// Manager name
        /// </summary>
        public string ManagerName => GetParam(MGR_PARAM_MGR_NAME, Environment.MachineName + "_Undefined-Manager");

        /// <summary>
        /// This will be true after the parameter have been successfully loaded from the database
        /// </summary>
        public bool ParamsLoadedFromDB { get; private set; }

        /// <summary>
        /// Dictionary of manager parameters
        /// </summary>
        public Dictionary<string, string> MgrParams { get; }

        #endregion

        #region "Events"

        /// <summary>
        /// Important error event (only raised if ParamsLoadedFromDB is false)
        /// </summary>
        public ErrorEventEventHandler CriticalErrorEvent;

        #endregion

        #region "Methods"

        /// <summary>
        /// Constructor
        /// </summary>
        public MgrSettings()
        {
            if (!LoadSettings())
            {
                if (string.Equals(ErrMsg, DEACTIVATED_LOCALLY))
                    throw new ApplicationException(DEACTIVATED_LOCALLY);
            ParamsLoadedFromDB = false;
            MgrParams = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

                throw new ApplicationException("Unable to initialize manager settings class: " + ErrMsg);
            }
        }

        /// <summary>
        /// Load manager settings from the config file
        /// </summary>
        /// <returns></returns>
        public bool LoadSettings()
        {
            var configFileSettings = LoadMgrSettingsFromFile();

            return LoadSettings(configFileSettings);
        }

        /// <summary>
        /// Updates manager settings, then loads settings from the database
        /// </summary>
        /// <param name="configFileSettings">Manager settings loaded from file AnalysisManagerProg.exe.config</param>
        /// <returns>True if successful; False on error</returns>
        /// <remarks></remarks>
        public bool LoadSettings(Dictionary<string, string> configFileSettings)
        {
            ErrMsg = string.Empty;

            MgrParams.Clear();

            foreach (var item in configFileSettings)
            {
                MgrParams.Add(item.Key, item.Value);
            }

            // Get directory for main executable
            var appPath = PRISM.FileProcessor.ProcessFilesOrDirectoriesBase.GetAppPath();
            var appFile = new FileInfo(appPath);
            SetParam("ApplicationPath", appFile.DirectoryName);

            // Test the settings retrieved from the config file
            if (!CheckInitialSettings(MgrParams))
            {
                // Error logging handled by CheckInitialSettings
                return false;
            }

            // Assure that MgrActive_Local is defined
            if (!MgrParams.TryGetValue(MGR_PARAM_MGR_ACTIVE_LOCAL, out _))
            {
                // MgrActive_Local parameter not defined defined in the AppName.exe.config file
                HandleParameterNotDefined(MGR_PARAM_MGR_ACTIVE_LOCAL);
            }

            // Get remaining settings from database
            if (!LoadMgrSettingsFromDB())
            {
                // Error logging handled by LoadMgrSettingsFromDB
                return false;
            }

            // Set flag indicating params have been loaded from the manager control database
            ParamsLoadedFromDB = true;

            // No problems found
            return true;
        }

        private Dictionary<string, string> LoadMgrSettingsFromFile()
        {
            // Load initial settings into string dictionary for return
            var mgrSettingsFromFile = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

            // Manager config db connection string
            var mgrCfgDBConnString = Properties.Settings.Default.MgrCnfgDbConnectStr;
            mgrSettingsFromFile.Add(MGR_PARAM_MGR_CFG_DB_CONN_STRING, mgrCfgDBConnString);

            // Manager active flag
            var mgrActiveLocal = Properties.Settings.Default.MgrActive_Local.ToString();
            mgrSettingsFromFile.Add(MGR_PARAM_MGR_ACTIVE_LOCAL, mgrActiveLocal);

            // Manager name
            var mgrName = Properties.Settings.Default.MgrName;
            if (mgrName.Contains("$ComputerName$"))
                mgrSettingsFromFile.Add(MGR_PARAM_MGR_NAME, mgrName.Replace("$ComputerName$", Environment.MachineName));
            else
                mgrSettingsFromFile.Add(MGR_PARAM_MGR_NAME, mgrName);

            // Default settings in use flag
            var usingDefaults = Properties.Settings.Default.UsingDefaults.ToString();
            mgrSettingsFromFile.Add(MGR_PARAM_USING_DEFAULTS, usingDefaults);

            return mgrSettingsFromFile;
        }


        /// <summary>
        /// Tests initial settings retrieved from config file
        /// </summary>
        /// <param name="paramDictionary"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        private bool CheckInitialSettings(IReadOnlyDictionary<string, string> paramDictionary)
        {
            // Verify manager settings dictionary exists
            if (paramDictionary == null)
            {
                ErrMsg = "CheckInitialSettings: Manager parameter string dictionary not found";
                OnErrorEvent(ErrMsg);
                return false;
            }

            if (!paramDictionary.TryGetValue(MGR_PARAM_USING_DEFAULTS, out var usingDefaultsText))
            {
                HandleParameterNotDefined(MGR_PARAM_USING_DEFAULTS);
            }
            else
            {
                if (bool.TryParse(usingDefaultsText, out var usingDefaults) && usingDefaults)
                {
                    ErrMsg = string.Format("MgrSettings.CheckInitialSettings; Config file problem, {0} contains UsingDefaults=True",
                                           GetConfigFileName());
                    ReportError(ErrMsg);
                    return false;
                }
            }

            // No problems found
            return true;
        }

        private string GetGroupNameFromSettings(DataTable dtSettings)
        {
            foreach (DataRow currentRow in dtSettings.Rows)
            {
                // Add the column heading and value to the dictionary
                var paramKey = DbCStr(currentRow[dtSettings.Columns["ParameterName"]]);

                if (string.Equals(paramKey, "MgrSettingGroupName", StringComparison.OrdinalIgnoreCase))
                {
                    var groupName = DbCStr(currentRow[dtSettings.Columns["ParameterValue"]]);
                    if (!string.IsNullOrWhiteSpace(groupName))
                    {
                        return groupName;
                    }

                    return string.Empty;
                }
            }

            return string.Empty;
        }

        private void HandleParameterNotDefined(string parameterName)
        {
            ErrMsg = string.Format("Parameter '{0}' is not defined defined in file {1}",
                                   parameterName,
                                   GetConfigFileName());
            ReportError(ErrMsg);
        }

        /// <summary>
        /// Gets manager config settings from manager control DB (Manager_Control)
        /// </summary>
        /// <returns>True if success, otherwise false</returns>
        /// <remarks>Performs retries if necessary.</remarks>
        private bool LoadMgrSettingsFromDB(bool logConnectionErrors = true)
        {

            var managerName = GetParam(MGR_PARAM_MGR_NAME, string.Empty);

            if (string.IsNullOrEmpty(managerName))
            {
                // MgrName parameter not defined defined in the AppName.exe.config file
                HandleParameterNotDefined(MGR_PARAM_MGR_NAME);
                return false;
            }

            var success = LoadMgrSettingsFromDBWork(managerName, out var mgrSettingsFromDB, logConnectionErrors, returnErrorIfNoParameters: true);
            if (!success)
            {
                return false;
            }

            success = StoreParameters(mgrSettingsFromDB, managerName, skipExistingParameters: false);

            while (success)
            {
                var mgrSettingsGroup = GetGroupNameFromSettings(mgrSettingsFromDB);
                if (string.IsNullOrEmpty(mgrSettingsGroup))
                {
                    break;
                }

                // This manager has group-based settings defined; load them now

                success = LoadMgrSettingsFromDBWork(mgrSettingsGroup, out var mgrGroupSettingsFromDB, logConnectionErrors, returnErrorIfNoParameters: false);

                if (success)
                {
                    success = StoreParameters(mgrGroupSettingsFromDB, mgrSettingsGroup, skipExistingParameters: true);
                }
            }

            return success;
        }

        /// <summary>
        /// Load manager settings from the database
        /// </summary>
        /// <param name="managerName">Manager name or manager group name</param>
        /// <param name="mgrSettingsFromDB">Output: manager settings</param>
        /// <param name="logConnectionErrors">When true, log connection errors</param>
        /// <param name="returnErrorIfNoParameters">When true, return an error if no parameters defined</param>
        /// <returns></returns>
        private bool LoadMgrSettingsFromDBWork(
            string managerName,
            out DataTable mgrSettingsFromDB,
            bool logConnectionErrors,
            bool returnErrorIfNoParameters)
        {

            mgrSettingsFromDB = null;

            var dbConnectionString = GetParam(MGR_PARAM_MGR_CFG_DB_CONN_STRING, "");

            if (string.IsNullOrEmpty(dbConnectionString))
            {
                // MgrCnfgDbConnectStr parameter not defined defined in the AppName.exe.config file
                HandleParameterNotDefined(MGR_PARAM_MGR_CFG_DB_CONN_STRING);
                return false;
            }

            var sqlQuery = "SELECT ParameterName, ParameterValue FROM V_MgrParams WHERE ManagerName = '" + managerName + "'";

            // Get a table holding the parameters for this manager
            var retryCount = 3;
            while (retryCount >= 0)
            {
                try
                {
                    using (var cn = new SqlConnection(dbConnectionString))
                    {
                        var cmd = new SqlCommand
                        {
                            CommandType = CommandType.Text,
                            CommandText = sqlQuery,
                            Connection = cn,
                            CommandTimeout = 30
                        };

                        using (var da = new SqlDataAdapter(cmd))
                        {
                            using (var ds = new DataSet())
                            {
                                da.Fill(ds);
                                mgrSettingsFromDB = ds.Tables[0];
                            }
                        }
                    }
                    break;
                }
                catch (Exception ex)
                {
                    retryCount -= 1;
                    var msg = string.Format("LoadMgrSettingsFromDB; Exception getting manager settings from database: {0}; " +
                                            "ConnectionString: {1}, RetryCount = {2}",
                                            ex.Message, dbConnectionString, retryCount);

                    if (logConnectionErrors)
                        ReportError(msg, criticalError: false);

                    // Delay for 5 seconds before trying again
                    if (retryCount >= 0)
                        ProgRunner.SleepMilliseconds(5000);
                }

            } // while

            // If loop exited due to errors, return false
            if (retryCount < 0)
            {
                // Log the message to the DB if the monthly Windows updates are not pending
                var criticalError = !WindowsUpdateStatus.ServerUpdatesArePending();

                ErrMsg = "MgrSettings.LoadMgrSettingsFromDB; Excessive failures attempting to retrieve manager settings from database for manager " + managerName;
                if (logConnectionErrors)
                    ReportError(ErrMsg, criticalError);

                return false;
            }

            // Validate that the data table object is initialized
            if (mgrSettingsFromDB == null)
            {
                // Data table not initialized
                ErrMsg = "LoadMgrSettingsFromDB; dtSettings data table is null; using " + dbConnectionString;
                if (logConnectionErrors)
                    ReportError(ErrMsg);

                return false;
            }

            // Verify at least one row returned
            if (mgrSettingsFromDB.Rows.Count < 1 && returnErrorIfNoParameters)
            {
                // Wrong number of rows returned
                ErrMsg = string.Format("MgrSettings.LoadMgrSettingsFromDB; Manager {0} not defined in the manager control database; using {1}",
                                       managerName, dbConnectionString);
                ReportError(ErrMsg);
                mgrSettingsFromDB.Dispose();
                return false;
            }

            return true;
        }

        /// <summary>
        /// Update mParamDictionary with settings in dtSettings, optionally skipping existing parameters
        /// </summary>
        /// <param name="dtSettings"></param>
        /// <param name="managerOrGroupName"></param>
        /// <param name="skipExistingParameters"></param>
        /// <returns></returns>
        private bool StoreParameters(DataTable dtSettings, string managerOrGroupName, bool skipExistingParameters)
        {
            bool success;

            try
            {
                foreach (DataRow currentRow in dtSettings.Rows)
                {
                    // Add the column heading and value to the dictionary
                    var paramKey = DbCStr(currentRow[dtSettings.Columns["ParameterName"]]);
                    var paramVal = DbCStr(currentRow[dtSettings.Columns["ParameterValue"]]);

                    if (paramKey.ToLower() == "perspective" && Environment.MachineName.ToLower().StartsWith("monroe"))
                    {
                        if (paramVal.ToLower() == "server")
                        {
                            paramVal = "client";
                            Console.WriteLine(
                                @"StoreParameters: Overriding manager perspective to be 'client' because impersonating a server-based manager from an office computer");
                        }
                    }

                    if (MgrParams.ContainsKey(paramKey))
                    {
                        if (!skipExistingParameters)
                        {
                            MgrParams[paramKey] = paramVal;
                        }
                    }
                    else
                    {
                        MgrParams.Add(paramKey, paramVal);
                    }
                }
                success = true;
            }
            catch (Exception ex)
            {
                ErrMsg = string.Format("MgrSettings.StoreParameters; Exception storing settings for manager '{0}': {1}",
                                       managerOrGroupName, ex.Message);
                ReportError(ErrMsg);
                success = false;
            }
            finally
            {
                dtSettings?.Dispose();
            }

            return success;
        }

        /// <summary>
        /// Gets a manager parameter
        /// </summary>
        /// <param name="itemKey"></param>
        /// <returns>Parameter value if found, otherwise empty string</returns>
        public string GetParam(string itemKey)
        {
            return GetParam(itemKey, string.Empty);
        }

        /// <summary>
        /// Gets a manager parameter
        /// </summary>
        /// <param name="itemKey">Parameter name</param>
        /// <param name="valueIfMissing">Value to return if the parameter does not exist</param>
        /// <returns>Parameter value if found, otherwise empty string</returns>
        public string GetParam(string itemKey, string valueIfMissing)
        {
            if (MgrParams.TryGetValue(itemKey, out var itemValue))
            {
                return itemValue ?? string.Empty;
            }

            return valueIfMissing ?? string.Empty;
        }

        /// <summary>
        /// Gets a manager parameter
        /// </summary>
        /// <param name="itemKey">Parameter name</param>
        /// <param name="valueIfMissing">Value to return if the parameter does not exist</param>
        /// <returns>Parameter value if found, otherwise empty string</returns>
        public bool GetParam(string itemKey, bool valueIfMissing)
        {
            if (MgrParams.TryGetValue(itemKey, out var valueText))
            {
                var value = UtilityMethods.CBoolSafe(valueText, valueIfMissing);
                return value;
            }

            return valueIfMissing;
        }

        /// <summary>
        /// Gets a manager parameter
        /// </summary>
        /// <param name="itemKey">Parameter name</param>
        /// <param name="valueIfMissing">Value to return if the parameter does not exist</param>
        /// <returns>Parameter value if found, otherwise empty string</returns>
        public int GetParam(string itemKey, int valueIfMissing)
        {
            if (MgrParams.TryGetValue(itemKey, out var valueText))
            {
                var value = UtilityMethods.CIntSafe(valueText, valueIfMissing);
                return value;
            }

            return valueIfMissing;
        }

        /// <summary>
        /// Set a manager parameter
        /// </summary>
        /// <param name="itemKey"></param>
        /// <param name="itemValue"></param>
        // ReSharper disable once UnusedMember.Global
        public void SetParam(string itemKey, string itemValue)
        {
            if (MgrParams.ContainsKey(itemKey))
            {
                MgrParams[itemKey] = itemValue;
            }
            else
            {
                MgrParams.Add(itemKey, itemValue);
            }
        }

        private string DbCStr(object inpObj)
        {
            if (inpObj == null)
            {
                return "";
            }
            return inpObj.ToString();
        }

        #endregion

        #region "Event Handlers"

        /// <summary>
        /// Report an important error
        /// </summary>
        /// <param name="message"></param>
        private void OnCriticalErrorEvent(string message)
        {
            if (CriticalErrorEvent == null && WriteToConsoleIfNoListener)
                ConsoleMsgUtils.ShowError(message, false, false, EmptyLinesBeforeErrorMessages);

            CriticalErrorEvent?.Invoke(message, null);
        }

        /// <summary>
        /// Raises a CriticalErrorEvent if criticalError is true and ParamsLoadedFromDB is false
        /// Otherwise, raises a normal error event
        /// </summary>
        /// <param name="errorMessage"></param>
        /// <param name="criticalError"></param>
        private void ReportError(string errorMessage, bool criticalError = true)
        {
            if (!ParamsLoadedFromDB && criticalError)
            {
                OnCriticalErrorEvent(errorMessage);
            }
            else
            {
                OnErrorEvent(errorMessage);
            }
        }

        #endregion
    }
}
